4-graceful-sigint/main.go

```diff
@@ -13,10 +13,38 @@
 
 package main
 
+import (
+	"context"
+	"os"
+	"os/signal"
+)
+
 func main() {
-	// Create a process
-	proc := MockProcess{}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	singnalFn := func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+		<-c
+		cancel()
+	}
 
-	// Run the process (blocking)
-	proc.Run()
+	go singnalFn()
+
+	proc := MockProcess{}
+	// Create a process
+	go func() {
+		// Run the process (blocking)
+		proc.Run()
+	} ()
+	
+	<-ctx.Done()
+	
+	go proc.Stop()
+	
+	ctx, cancel = context.WithCancel(context.Background())
+	
+	go singnalFn()
```

Comment:
not sure if you need to do it again
I assume signal.Notify would continue to send notifications on your channel so you could just listen on it instead of context